src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\EventListener;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpKernel\Event\FinishRequestEvent;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+use Symfony\Component\Translation\LocaleAwareInterface;
+use TypeError;
+
+/**
+ * Inject request locale to each services that implements LocaleAwareInterface
+ *
+ * @author Pierre Bobiet <pierrebobiet@gmail.com>
+ */
+class LocaleAwareListener implements EventSubscriberInterface
+{
+    /**
+     * @var \Symfony\Component\Translation\LocaleAwareInterface[]
+     */
+    private $localeAwareServices;
+    private $requestStack;
+
+    public function __construct(\Traversable $localeAwareServices, RequestStack $requestStack)
+    {
+        $this->setLocaleAwareServices($localeAwareServices);
+        $this->requestStack = $requestStack;
+    }
+
+    private function setLocaleAwareServices(\Traversable $localeAwareServices)
+    {
+        foreach ($localeAwareServices as $i => $service) {
```

Comment:
This means you will lose the laziness of iterable arguments, and force instantiation of all locale aware services as soon as the listener is instantiated.

Reply:
Maybe the check could be done in a compiler pass like it's done for voters.